src/reducers/index.ts

```diff
@@ -52,21 +52,51 @@ function checkStoreInvariants(state: StoreState) {
     state.currentMatchIndex >= -1 &&
       state.currentMatchIndex <= state.matchesList.length
   );
+
+  checkCondition(
+    'every matchId in matchIdToMatchState is also present in matchesList',
+    Object.keys(state.matchIdToMatchState).reduce(
+      (a, e) =>
+        a && state.matchesList.filter(v => v.matchId === e).length === 1,
+      true
+    )
+  );
 }
 
 function reduce(state: StoreState, action: Action) {
   if (action.setGamesList) {
     return { ...state, gamesList: action.setGamesList };
   } else if (action.setMatchesList) {
-    return { ...state, matchesList: action.setMatchesList };
+    let { matchesList, matchIdToMatchState, ...rest } = state;
+    let newMatchIdToMatchState = {};
+    action.setMatchesList.forEach(e => {
+      if (e.matchId in matchIdToMatchState) {
+        newMatchIdToMatchState[e.matchId] = matchIdToMatchState[e.matchId];
+      } else {
+        newMatchIdToMatchState[e.matchId] = {};
```

Comment:
delete this line.
matchIdToMatchState[someID] should tell you if we have state or not.
Setting {} is confusing :)